RFC: Demote i686-pc-windows-gnu to Tier 2

Noratrieb opened the pull request:
[Rendered](https://github.com/Noratrieb/rfcs/blob/i686-pc-windows-gnu/text/3771-demote-i686-pc-windows-gnu.md)

ChrisDenton commented on text/3771-demote-i686-pc-windows-gnu.md:
```diff
@@ -0,0 +1,172 @@
+
+## Background
+
+Rust has supported Windows ever since before 1.0, with two different flavors of Windows targets: MSVC-based and GNU-based.
+MSVC-based targets (for example the main Windows target `x86_64-pc-windows-msvc`) use Microsoft's native `link.exe` linker and libraries, while GNU-based targets (like `i686-pc-windows-gnu`) are built entirely from free software components like `gcc`, `ld`, and MinGW.
```
Nitpick: technically MSVC wasn't supported until, I think, 1.2. Or at least not as tier 1 or 2.

mati865 commented on text/3771-demote-i686-pc-windows-gnu.md:
```diff
@@ -0,0 +1,172 @@
+MSYS2, a major distributor of the GNU-based Windows platform, has been [dropping some 32-bit packages](https://www.msys2.org/news/#2023-12-13-starting-to-drop-some-32-bit-packages) and [no longer distributes Clang for 32-bit](https://github.com/msys2/MINGW-packages/pull/21998), showing even their shift away from the platform.
+In response to inquiries about their opinion on reducing support for the target, [MSYS2 folks were positive](https://github.com/msys2/MINGW-packages/issues/23346).
+
+[^1]: despite saying he is "not a maintainer", mati865 is effectively our maintainer for x86_64-pc-windows-gnu and x86_64-pc-windows-gnullvm https://rust-lang.zulipchat.com/#narrow/channel/233931-t-compiler.2Fmajor-changes/topic/Demote.20.60i686-pc-windows-gnu.60.20compiler-team.23822/near/490675824
```
I do maintain `*-gnullvm` targets though 😉

jieyouxu commented on text/3771-demote-i686-pc-windows-gnu.md:
```diff
@@ -0,0 +1,172 @@
+
+### 32-bit x86 Problems
+
+While some of these issues apply to all GNU-based targets, 32-bit x86 seems to be especially affected.
+And when a 32-bit Windows GNU issue comes up, contributors rarely actually investigate it, because it is such a complex and nonstandard environment compared to 64-bit Windows GNU, which is a lot easier to set up and work with.
```
Remark: one such example: https://github.com/rust-lang/rust/issues/136795

jieyouxu commented on text/3771-demote-i686-pc-windows-gnu.md:
```diff
@@ -0,0 +1,172 @@
+
+As mentioned above, there are issues and it does cause a fair share of problems.
+
+> The target must not disable an excessive number of tests or pieces of tests in the testsuite in order to do so. This is a subjective requirement.
+
+A fair amount of tests are disabled with an open issue with no comments.
+I would say that it is on the edge of being excessive, not quite having reached that amount (but it is likely that will be reached eventually).
```
Remark:

- @saethlin did a lot of work in https://github.com/rust-lang/rust/pull/134777 to remove `//@ ignore-windows` / `//@ ignore-gnu` / `//@ ignore-windows-gnu` or to tighten the ignores. There's a bunch of them because compiler contributors tend to give up debugging why a test fails in `i686-mingw` (or less likely `x86_64-mingw-{1,2}`) but not on `x86_64-msvc` (and `i686-msvc`) and just slap an `ignore-windows` or `ignore-windows-gnu` (which also blanket ignores the 64-bit windows-gnu target).
- Also an example: in https://github.com/rust-lang/rust/pull/135572 I did not bother with trying to make the test work on windows-gnu at all (let alone 32-bit windows-gnu) because it's too annoying to setup the environment and also debug why tests may be different in the context of windows-gnu specifically. I have access to native x86_64 msvc, that I'm willing to make the test pass on.

jieyouxu commented on text/3771-demote-i686-pc-windows-gnu.md:
```diff
@@ -0,0 +1,172 @@
+While this cannot be quantified precisely, the download counts suggest that this target is less popular than some other Tier 2 targets like FreeBSD.
+Therefore, we are going to treat this as false.
+
+> The target maintainer team must include at least 3 developers.
+
+This is not the case at all. There is currently no maintainer team.
+Though we should note that this is currently also true for many other Tier 1 targets, as this is a new rule not upheld everywhere yet.
+But experience tells that it is highly unlikely that 3 maintainers for 32 bit Windows GNU will be found.
```
Remark: and I think it's arguably misleading to say `i686-pc-windows-gnu` is a Tier 1 target when it does not really receive Tier 1 maintenance support IMHO.

jieyouxu commented on text/3771-demote-i686-pc-windows-gnu.md:
```diff
@@ -0,0 +1,172 @@
+
+## Conclusion
+
+Given the usage count and lack of maintenance leading to more than one requirement not being fulfilled, it becomes clear that `i686-pc-windows-gnu` is not worthy of being a Tier 1 target and is already getting much worse support than expected from a Tier 1 target.
```
Remark: arguably, it also imposes a maintenance burden on compiler contributors because if a test fails on just `i686-mingw` (32-bit windows-gnu) but not 64-bit windows-{msvc,gnu} it's very annoying to bless that.

jieyouxu commented on text/3771-demote-i686-pc-windows-gnu.md:
```diff
@@ -0,0 +1,172 @@
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+`x86_64-pc-windows-gnu` will remain a Tier 1 target after this RFC.
+While its popularity is more aligned with Tier 1, it suffers from the same lack of maintenance (but to a lesser degree) as its 32-bit cousin.
+It may be demoted as well in the future.
```
Remark: at least the 64-bit windows-gnu is more accessible as a *host* target, unlike the 32-bit windows-gnu.

programmerjake replied:
if you want to cross-compile from linux to windows, the 64-bit windows-gnu (or gnullvm) target is the best way to do that, so I expect it to stay around a lot longer. we'd probably want to add an arm gnu target too at some point.

workingjubilee replied:
I expect that we would have an aarch64-pc-windows-gnullvm or whatever target instead.

mati865 replied:
This discussion is drifting away from the culprit, but I don't see value in ARM targets (32-bit).

AArch64 is already there as a cross compilation target (gnullvm only until GNU catches up), and once LLVM 20 goes live, I'll try to start MCP for promoting 64-bit gnullvm targets to hosts.

programmerjake replied:
> AArch64 is already there as a cross compilation target (gnullvm only until GNU catches up), and once LLVM 20 goes live, I'll try to start MCP for promoting 64-bit gnullvm targets to hosts.

nice! I had meant aarch64, but iirc windows doesn't call it that for some reason...

jieyouxu reviewed (commented):
I'm in favor of this, left some remarks / additional context which are not meant to be requested changes.

Noratrieb commented:
I addressed some of the feedback I thought was worth putting into the RFC text itself.

I don't expect this to be controversial, so
@rfcbot fcp merge

programmerjake commented on text/3771-demote-i686-pc-windows-gnu.md:
```diff
@@ -0,0 +1,174 @@
+
+A fair amount of tests are disabled with an open issue with no comments.
+I would say that it is on the edge of being excessive, not quite having reached that amount (but it is likely that will be reached eventually).
+For example, [#134777](https://github.com/rust-lang/rust/pull/134777) observed and un-ignored a lot of ignored Winodws tests, many of which were likely ignored on all of Windows because of Winodws GNU issues.
```
```suggestion
For example, [#134777](https://github.com/rust-lang/rust/pull/134777) observed and un-ignored a lot of ignored Windows tests, many of which were likely ignored on all of Windows because of Windows GNU issues.
```

Noratrieb commented:
ping @Mark-Simulacrum @SparrowLii @cjgillot @estebank @petrochenkov @spastorino

joshtriplett commented on text/3771-demote-i686-pc-windows-gnu.md:
```diff
@@ -0,0 +1,174 @@
+That demotion will not need an RFC.
+
+The `*-windows-gnullvm` targets, which are based on LLVM instead of GNU tools, may see increased maintenance and popularity in the future, replacing the `*-windows-gnu` targets.
+But it seems unlikely that `i686-pc-windows-gnullvm` would ever acquire Tier 1 status.
```
```suggestion
(However, it is unclear whether `i686-pc-windows-gnullvm` would get enough interest and support to attain Tier 1 status.)
```

mati865 replied:
It's straight up impossible.
LLVM uses memory mapping for many IO operations. On 32-bit Windows, it can not even host itself due to insufficient address space.

joshtriplett replied:
@mati865 Is that an issue with the compiler or with the linker? (Would it work if linked with a different linker?)

In any case, thanks for the clarification. I'll edit the suggestion.

Noratrieb commented:
Thanks everyone! I have created https://github.com/rust-lang/rust/issues/138422 to track the necessary steps.
